File: Problem 4/solution.cpp
/*
This program finds the largest palindrome made from the 
product of two 3-digit numbers.
*/

#include <iostream>
#include <math.h>

bool is_palindrome(int n)
{
    /*
        This function checks if a number is a Palindrome
        or not.
    */
    int result = 0;
    int temp = n;
    // reverse the number
    while (temp > 0)
    {
        result *= 10;
        result += temp % 10;
        temp /= 10;
        temp = floor(temp);
    }
    // console.log(result);
    return result == n;
}

int main()
{
    int largest = 0;
    int product = 0;
    for (int i = 100; i < 1000; i++)
    {
        for (int j = 100; j < 1000; j++)
        {
            product = i * j;
            if (is_palindrome(product) && largest < product)
                largest = product;
        }
    }
    std::cout << largest << std::endl;
}

File: Problem 4/solution.py
"""
This program finds the largest palindrome made from the 
product of two 3-digit numbers.
"""


def is_palindrome(n) -> bool:
    """
        This function checks if a number is a Palindrome
        or not.
    """
    result = 0
    temp = n
    # reverse the number
    while temp > 0:
        result *= 10
        result += temp % 10
        temp //= 10
    return result == n


largest = 0
for i in range(100, 1000):
    for j in range(100, 1000):
        product = i * j
        if is_palindrome(product) and largest < product:
            largest = product
# print(i,j)
print(largest)
